### PR TITLE
fix issue #611

### LINKED ISSUE #611: Malformed custom targets Install_... Build run out of memory in docker-based CI
When building Radium in memory-bounded environment (e.g. free Gitlab docker runners, limited to 2gb of memory)
the command `make Install_` used to compile configured components and install them in one command fail.

This command, as defined in the main `CMakeLists.txt`, line 124, launches as many compilation process as there are files to compile. 
This makes `gcc` to run out of memory as compiling Radium this way requires 16GB of memory (on Linux) due to template instantiation.

The way the target `Install_` is defined prevent the caller to set the max number of jobs for its compilation environment  (e.g. by using `make -j 4 Install_`) as the "infinite" number of job to use is set directly in the target definition.

This malformed target definition is also present in the file `cmake/Config.cmake.in`, line 6 and generates the same effect (run out of memory) when compiling some Radium client libraries or apps that rely on Radium's templated classes.

The solution to this is to remove the `-j` option in the definition of the custom target and to let the user, either to 
  - explicitly set the option when calling make (e.g. `make -j <jobs> Install_`) or by using `cmake --build <dir> -j <jobs> --target Install_` 
  - or define the cmake environment variable `CMAKE_BUILD_PARALLEL_LEVEL` (see https://cmake.org/cmake/help/v3.16/envvar/CMAKE_BUILD_PARALLEL_LEVEL.html) before compiling using `cmake --build ...`

Note that if Radium is compiled and installed using 2 commands (i.e. `make && make Install_`) the problem can be avoided (but is still there) but this is not documented. This allow to compile in low memory environment but spawning an infinite number of threads for only installing everything is then useless.

### PR DESCRIPTION
**Pull request by MathiasPaulin**
fix #611

### CODE REVIEW

**Comment by nmellado** [comment]
Thanks.
One alternative would be to modify the name of `RADIUM_BUILD_EXTERNAL_PARALLEL_LEVEL` and use it to also control thE number of installation thread.

**Comment by MathiasPaulin** [comment]
> Thanks.
> One alternative would be to modify the name of `RADIUM_BUILD_EXTERNAL_PARALLEL_LEVEL` and use it to also control thE number of installation thread.

I agree, `RADIUM_BUILD_EXTERNAL_PARALLEL_LEVEL` is redondant with the cmake environment variable  `CMAKE_BUILD_PARALLEL_LEVEL`. But I'm not sure that forcing  (by hiding the parallel configuration in the Radium cmake scripts) parallel build or parallel install is a good thing. I rather prefer to let the user specify this. When the environment variable is set, all the build using `cmake --build ...` will be parallelized (even with ninja build). Letting the user calling `make -j <jobs> ...` also allow to have finer control on the build process. If the option `-j <jobs>`is set in a cmake scripts, the user can't change it when calling make or build.

**Comment by nmellado** [comment]
It os not redundant as it is just a way to compute a value that will be used to compile the sub modules.
The only thing that change is that we set the default as « all cores » as in cmake it is « one core ».
We can change that, but internally we still need that variable to propagate this parameter to externals.

**Comment by MathiasPaulin** [comment]
The definition of `RADIUM_BUILD_EXTERNAL_PARALLEL_LEVEL` is redundant as environment variable have global scope and the value os such variables is the one from the calling process (https://cmake.org/cmake/help/v3.15/manual/cmake-language.7.html#cmake-language-environment-variables). I agree you just want to change the default from `1` to `N` for building externals and we can do that using `CMAKE_BUILD_PARALLEL_LEVEL`. But this can stay as it is (until a problem similar to the one exposed in the issue #611 is met when building externals.
What causes a problem is the `-j` option defined in the custom target `Install_`. It asks for "infinite" number of job for the processing of this target (and its dependencies) and prevent the user to change the number of jobs used for this target. Whereas, setting `CMAKE_BUILD_PARALLEL_LEVEL` allow to change this number of job at each build operation while setting the default.

What we can do is to set the variable to the number of core if the user (or the calling process) did not set it so that, by default, all is built using `N` threads and the user still have the option to change this number of threads dynamicaly. Either by setting the `CMAKE_BUILD_PARALLEL_LEVEL` as a system environment variable of by adding `-j P` to any `cmake --build ...` of `make ...` command. 

The last commit implement this and I've verified. When the `CMAKE_BUILD_PARALLEL_LEVEL` is not set by the user in its environment, all is built using NCores threads. When the user set this variable, its value is used. Whatever the case, calling `cmake --build -j P` or `make -j P` changes the number of threads used for the current command.

File: cmake/Config.cmake.in [hunk 1]
```diff
@@ -1,13 +1,8 @@
...
 # Custom Install target, used in run target in source/CMakeLists.txt
-if (CMAKE_GENERATOR MATCHES "Makefiles")
-    # Make it multithreaded
-    add_custom_target( Install_${PROJECT_NAME} COMMAND "${CMAKE_COMMAND}" --build . --target
-            install -- -j WORKING_DIRECTORY "${CMAKE_BINARY_DIR}")
-else()
-    add_custom_target( Install_${PROJECT_NAME} COMMAND "${CMAKE_COMMAND}" --build . --target
-            install WORKING_DIRECTORY "${CMAKE_BINARY_DIR}")
-endif()
+add_custom_target( Install_${PROJECT_NAME} COMMAND "${CMAKE_COMMAND}" --build . --target
```
Comment (dlyr) [inline cmake/Config.cmake.in:3]: don't we want o have `--parallel` here ?  [see doc](https://cmake.org/cmake/help/v3.18/manual/cmake.1.html#build-a-project)

Comment (dlyr) [reply cmake/Config.cmake.in:3]: Well, I'm not sure of the doc... 
so if no -parallel, the use env var ? or if no number given then use env var ...

Comment (dlyr) [reply cmake/Config.cmake.in:3]: Ok, I confirm, without -j or --parallel, the --build use the number of threads from env var.

**Review by dlyr — APPROVED** [review]

**Pull request merged**